models/tiegcm/model_mod.f90

```diff
@@ -769,1225 +509,345 @@ end subroutine get_state_meta_data
+                                            domain_id(SECONDARY_DOM), ivarZG)
+         height = get_state(height_idx, state_handle)/100.0_r8
+   
+      case ('lev') ! height on midpoint
+        height_idx = get_dart_vector_index(lon_index, lat_index, lev_index, &
+                                   domain_id(SECONDARY_DOM), ivarZG)
+        height1 = get_state(height_idx, state_handle)/100.0_r8
+        height_idx = get_dart_vector_index(lon_index, lat_index, lev_index+1, &
+                           domain_id(SECONDARY_DOM), ivarZG)
+        height2 = get_state(height_idx, state_handle)/100.0_r8
+        height = (height1 + height2) / 2.0_r8
+   
+      case default
+       call error_handler(E_ERR, 'convert_vertical_state', 'expecting ilev or ilat dimension')
+   end select
+   
+   locs(i) = set_location(lons(lon_index), lats(lat_index), height(1), VERTISHEIGHT)
 
-!-------------------------------------------------------------------------------
+end do
 
+end subroutine convert_vertical_state
 
-function nc_write_model_vars( ncid, statevec, copyindex, timeindex ) result (ierr)
-! TJH 24 Oct 2006 -- Writes the model variables to a netCDF file.
-!
-! TJH 29 Jul 2003 -- for the moment, all errors are fatal, so the
-! return code is always '0 == normal', since the fatal errors stop execution.
-!
-! assim_model_mod:init_diag_output uses information from the location_mod
-!     to define the location dimension and variable ID. All we need to do
-!     is query, verify, and fill ...
+!-------------------------------------------------------------------------------
 
-integer,                intent(in) :: ncid
-real(r8), dimension(:), intent(in) :: statevec
-integer,                intent(in) :: copyindex
-integer,                intent(in) :: timeindex
-integer                            :: ierr ! return value of function
+function read_model_time(filename, ncfileid, lasttime)
+! Gets the latest time in the netCDF file.
+character(len=*),  intent(in) :: filename
+integer, optional, intent(in) :: ncfileid
+integer, optional, intent(in) :: lasttime
+type(time_type) :: read_model_time
 
-integer :: nDimensions, nVariables, nAttributes, unlimitedDimID
-integer :: StateVarID
+integer :: ncid, TimeDimID, time_dimlen, DimID, dimlen, VarID
 
-integer, dimension(NF90_MAX_VAR_DIMS) :: dimIDs, mystart, mycount
-character(len=NF90_MAX_NAME)          :: varname
-character(len=NF90_MAX_NAME),dimension(NF90_MAX_VAR_DIMS) :: dimnames
-integer :: i, ivar, VarID, ncNdims, dimlen
-integer :: TimeDimID, CopyDimID
+integer,  parameter         :: nmtime = 3
+integer,  dimension(nmtime) :: mtime  ! day, hour, minute
+integer                     :: year, doy, utsec
+integer, allocatable, dimension(:,:) :: mtimetmp
+integer, allocatable, dimension(:)   :: yeartmp
 
-if ( .not. module_initialized ) call static_init_model
+if ( present(ncfileid) ) then
+   ncid = ncfileid
+else
+   call nc_check(nf90_open(filename, NF90_NOWRITE, ncid), &
```

Comment:
eventually these calls could use the dart netcdf utility routines.

Reply:
I'll do this in this pull request - good time to do it.

Reply:
done [cb492ec](https://github.com/NCAR/DART/commit/cb492ecf7a6469105de176a65a27660589180cde)